### DISCUSSION TITLE
Ways of styling JS

### DISCUSSION THREAD

**Discussion by PaulFarault** [TOSIT-IO/TDP#52] [Ideas]
We have to choose how we are going to style our React components on `tdp-ui`. I see 2 main ways to do so:

- Component libraries
- Utility classes

## Component library

The big **advantage** of component library is that they provide ready-to-use components.

Component libraries are great for some use cases but can be a **disadvantage** for more specific needs. Each library is very opinionated and will require specific learning from users. Customizing default components can quickly become complex.

The use of component libraries is possible for certain components, it will however be necessary to complete it with other tools (utility classes, css-in-js, css modules, etc.). Many good libraries exist, we can discuss selection criteria if we go with this option:

- [React Bootstrap](https://react-bootstrap.github.io)
- [Mantine](https://mantine.dev/)
- [Ant Design](https://ant.design/)
- [Material UI](https://mui.com/)
- [Rebass](https://rebassjs.org/)
- [Chakra UI](https://chakra-ui.com/)
- [Tamagui](https://tamagui.dev/)
- [React Suite](https://rsuitejs.com/)
- [Prime React](https://primefaces.org/primereact/)
- ~[Grommet](https://v2.grommet.io/)~ (mobile oriented)
- [Semantic UI React](https://react.semantic-ui.com/)
- [Blueprint](https://blueprintjs.com/)

## Utility class

The main **advantage** of utility classes is their ease of use, by placing the attributes directly in the JSX. The classes are self-explanatory enough for beginners to work with. They make it possible to create protean solutions.

**Disadvantages** of utility classes is that they can quickly complicate the JSX with lengthy `className`. This problem can be solved by enforcing strong architectural rules. They also don't provide ready-to-use component, everything is to be made.

The go-to utility class today is [Tailwind CSS](https://tailwindcss.com/). It also benefits from a library of premade components from which we can draw inspiration (see [Tailwind UI](https://tailwindui.com/)).

We can also benefit from both of those world by using a component library based on Tailwind CSS such as [DaisyUI](https://daisyui.com/).

**Comment by rpignolet** [comment]
I wonder how you can configure theme with Tailwind. I look at https://tailwindui.com/components/preview and I see this HTML code for buttons:
``` html
<div class="flex flex-1 justify-between sm:hidden">
  <a href="#" class="relative inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50">Previous</a>
  <a href="#" class="relative ml-3 inline-flex items-center rounded-md border border-gray-300 bg-white px-4 py-2 text-sm font-medium text-gray-700 hover:bg-gray-50">Next</a>
</div>
```
I see `bg-white` for example so I suppose this will draw a white background for the button. How can I customize the button color without changing the HTML code ? It seems impossible to customize theme.
Moreover I do not see how we can maintain the same button design for all our buttons if we must copy and paste so much classes for each button. If we do a button component we can avoid copy and paste but in the end we will make ourselves a component library and we want to avoid that IMO.

I look at https://mui.com/material-ui/customization/theming/ to get an example how to configure theme and this does not require changing HTML. Am I wrong ?

**Reply by gboutry** [reply]
From tailwind documentation:
- https://tailwindcss.com/docs/theme
- https://tailwindcss.com/docs/presets

**Reply by rpignolet** [reply]
Yes but it confirm what I said. This is just to define default theme but not override what you write in the HTML. If you use a `bg-white` on the button, how I override the color to `red` for example without changing the `bg-white` in the HTML ?

**Reply by gboutry** [reply]
You do not use `bg-white` but `bg-primary` for example. The goal of tailwind is to generate utility classes from your configuration. You can create your custom colors like `primary` and have presets to change its value.

https://tailwindcss.com/docs/customizing-colors#using-css-variables

**Reply by PaulFarault** [reply]
I see tailwind more as a way to simplify writing CSS. It surely doesn't have the same capacities of a full component library.

I should have put it in my post but we can use a good compromise: a component library based on Tailwind. I am thinking in particular of [DaisyUI](https://daisyui.com/) which is very promising (theming system, basic components, works with Tailwind).

**Reply by wdavidw** [reply]
Working with something like `bg-primary` is not very flexible. When more flexibility is required, it is also possible to [change the default color palette](https://tailwindcss.com/docs/customizing-colors#color-object-syntax), for example `bg-slate-200` can be set to a slightly bluish or greenish grey instead of the default.

**Reply by gboutry** [reply]
But, if we want the color to be customizable via theme selection, you want to write everywhere:
```javascript
clsx({
'bg-white': color === 'white',
'bg-red': color === 'red',
})
```

This does not seem very flexible either.

How would you implement multiple theme with tailwindcss and react ?

**Reply by rpignolet** [reply]
Using `bg-primary` is flexible because you design a theme with primary and secondary colors. This is explain why here https://material.io/design/color/the-color-system.html#color-usage-and-palettes for the example. You do not design a button to be `gray`, you design a button to be a primary action, secondary action, etc. and after, a designer decide how the color for primary action should be.

**Reply by wdavidw** [reply]
In my case, it was not about theme but for a button componennt. I have a default button component, eg with background green, which I wish to extend/recycle with an important button component, eg red.

I don't think we want to support multiple themes. It is more interesting to customize the default theme in both light and dark modes. This is well documented in the [theme doc](https://tailwindcss.com/docs/theme). 

To support multiple theme, my current approach is to work with typography:

```js
module.exports = {
  ...,
  plugins: [
    require('@tailwindcss/typography'),
  ],
  theme: {
    extend: {
      ...,
      typography: (theme) => ({
        invert: {
          css: {
            a: {
              color: theme("colors.sky.200"),
              "&:hover": {
                color: theme("colors.sky.400"),
              },
            },
            ...
          },
        },
        DEFAULT: {
          css: {
            a: {
              color: theme("colors.sky.800"),
              "&:hover": {
                color: theme("colors.sky.600"),
              },
            },
            ...
}}})}}}
```

Here, we have the default prose, as well as invert, but we could have additional custom prefixes.

**Reply by gboutry** [reply]
> I don't think we want to support multiple themes. It is more interesting to customize the default theme in both light and dark modes. This is well documented in the [theme doc](https://tailwindcss.com/docs/theme).

The goal of multi-theming is to be able to set different colors depending the environment (dev, prod...).

Are you saying having this is a bad idea ?

Or maybe you meant (in last group meeting), that we should just change the color of the band on the left and not a global theme ?

**Reply by wdavidw** [reply]
More like the last one. I covered theming in a general context and not for the environment customization. Theming is a much larger/complex topic than just providing some color guidance relative to the environment. It affects all the components. Our environment requirement is very localized inside the UI, for example in the header. To address the environment personalization, we may:

- extends the default theme, eg in tailwind.config.js, but this is kind of static and global to the all page
- provide personalization options from a config file, or a rest endpoint, eg change the header bg color and some text, it could be configurable from the admin area later one

Some easy customization inside the header seems enough, don't you agree:

```js
export default () => (
  <header className="flex">
    <Logo />
    <p className={clsx(
      'grow', {
        [`bg-[${settings.admin_bgcolor}]`]: !!settings.admin_bgcolor
      }
    )}>
      {settings.admin_label || 'TDP UI'}
    </p>
    <Menu />
  <header>
)
```

It is possible to support multiple themes in Tailwind, with the help of plugins but I don't think we need to go that route for now.

**Comment by wdavidw** [comment]
I have worked intensively with Bootstrap, Semantic UI, vanilla CSS styles, Material UI and Tailwind. Here is my feedback. 

My favorites are Material UI and Tailwind. Since they cover the two approaches, I focus on them.

Whichever the choices, there are always some limit to the framework in use:

- advanced personalization
- custom components not available on the shelves

Rich toolsets like Material UI comes with a very large set of components and it is always possible to compose new ones. Advanced personalization is more problematic as you often end up in the source code with the need to understand the source in order to personalize its behavior. There are some strong adherence with the framework which weaken the ability to transparently support upcoming versions.

I don't think there is a lot to say about Material UI. The approach is quite common, the library is mature, extensive and well designed. There is more confusion with Tailwind. I was personally reluctant to use it for a few years. Until I tried, and it feels smart.

Tailwind provide more control because it works at a lower level, HTML/CSS, and comes without any components. However, the approach has proved to be very pleasant, powerful and even productive. The developer is in control and can compose his own set of components, which in the React world is not such a constraint. Tailwind CSS is also extended with a quality eco-system among which Tailwind UI, Heroicons, Headless UI, Flowbite, daisyUI, ...

About the className argument mentioned above, it is true. There is a lot of classes to integrate. While a bit scared at first, it actually makes a lot of sense and is also an advantage. With the usage of tools like `clsx`, they are easy to read. Classes usage and names are natural. The creation of components makes them easy to be reused, for example to create a form library. And above all, this is the condition to feel in control of the code. The productivity with Tailwind, at least in my case, is greater than with any other library.

The example below illustrate how an input component is styled, it is extensible with the `className` props and it supports focus, dark mode and disable state:

```jsx
    <input
      className={clsx(
        className,
        'mt-1 block w-full rounded-md shadow-sm sm:text-sm',
        'dark:bg-slate-100/10',
        'border-gray-300 focus:border-indigo-500 focus:ring-indigo-500',
        'disabled:opacity-30', {
        'bg-red-500': color === 'red',
      })}
      {...attributes}
    />
```

Over the last few months using it, I found one bug, mentioned in the comments above. It finds its origin in the HTML specification. There is not ordering preference in classes. You mean that `class="bg-red bg-white"` renders the same as `class="bg-white bg-red"` and you have no control over it. There are some ways to mitigate the issue, using `!` for important (oh no), activating some libraries which analyses the classes and injecting a custom property like I do with `color` above. I only encountered the issue on rare occasions, and this is the only issue I had. Compared to the hours spent on the Material UI source (which by now I know very well), this is acceptable to me.

**Comment by PaulFarault** [comment]
I may have introduced the discussion wrong. Comparing Tailwind CSS and MUI is equivalent to React and Vue: we can't.

IMO there are 2 questions in this discussion:

1. **How do we want to style React?**   
  Next.js natively supports [many ways](https://nextjs.org/docs/basic-features/built-in-css-support) of handling CSS (global, module, sass, CSS-in-JS), adding Tailwind CSS could be beneficial as previous comments show.
2. **Should we use a component library in addition? If yes, which one ?**   
  It could be MUI, DaisyUI (which is based on Tailwind CSS) or others.

**Reply by rpignolet** [reply]
Yes you wrote your message before me.

We should compare for example, [Material UI](https://mui.com/) and [DaisyUI](https://daisyui.com/). We can compare others, this is just a example. I'm not against plugin for Tailwind, just using it directly. It is like comparing React and Vue.js which does not meet the same need that's why we ended up comparing Next.js and Vue.js.

Tailwind is great to make a design from scratch but we do not want to make a new component library, we are not designer, we want to dev a UI with a component ready lib.

**Reply by wdavidw** [reply]
If we go the Tailwind road, the majority of what we need is already in [Tailwind UI](https://tailwindui.com/components).

**Comment by PaulFarault** [comment]
From the list I posted, we can already filter the [small libraries](https://npmtrends.com/@blueprintjs/core-vs-@chakra-ui/react-vs-@mantine/core-vs-@material-ui/core-vs-@mui/material-vs-antd-vs-daisyui-vs-primereact-vs-react-bootstrap-vs-rebass) and keep only the [5 most popular](https://npmtrends.com/@chakra-ui/react-vs-@material-ui/core-vs-@mui/material-vs-antd-vs-react-bootstrap):

1. [MUI](https://mui.com/) ([`@material-ui/core`](https://www.npmjs.com/package/@material-ui/core)), **1,9M** weekly downloads
2. [MUI](https://mui.com/) ([`@mui`](https://www.npmjs.com/package/@mui/material)), **1,8M** weekly downloads
3. [React-Bootstrap](https://react-bootstrap.github.io/) ([`react-bootstrap`](https://www.npmjs.com/package/react-bootstrap)), **1,2M** weekly downloads
4. [Ant Design](https://ant.design/) ([`antd`](https://www.npmjs.com/package/antd)), **0,86M** weekly downloads
5. [Chakra UI](https://chakra-ui.com) ([`@chakra-ui/react`](https://www.npmjs.com/package/@chakra-ui/react)), **0,38M** weekly downloads

By far the most popular library is MUI. The top 2 packages from the list are from them, due to [a recent change](https://mui.com/blog/material-ui-is-now-mui/) in the branding of the company. Material UI is the implementation of [Material Design](https://material.io/design), the Google design system, one of the most robust and accessible.

Among the benefits of MUI :

- Biggest number of users
- Well maintained
- Lot of components
- [Pre-made templates](https://mui.com/material-ui/getting-started/templates/)
- [Compatibility with Tailwind CSS](https://mui.com/base/guides/working-with-tailwind-css/)
- [Theme builder](https://bareynol.github.io/mui-theme-creator/) (community)

The other libraries in this top 5 also offer a lot of components and are quite robust too. We can mention that: 

- React-Bootsrap is based on... [Boostrap](https://getbootstrap.com/), which was the go-to for many years and still has a large user base.
- Chakra UI seems to be the more flexible, with components that extend easily (see the design [principles](https://chakra-ui.com/getting-started/principles)).
- Ant Design traction seems to be concentrated in China (see [users](https://ant.design/docs/react/introduce#Companies-using-antd) and [Google trends](https://trends.google.com/trends/explore?q=ant%20design,material%20ui)). The home page even contains articles that are not translated.

IMO if we choose to use a component library, we can't be wrong with MUI.

**Reply by PaulFarault** [reply]
To follow up this discussion and my [PR](https://github.com/TOSIT-IO/tdp-ui/pull/7) using the Paperbase template (MUI), which was found too complicated, I made 3 POCs to help us decide:

- [Option 1](https://github.com/TOSIT-IO/tdp-ui/tree/6-mui-dashboard-template): basic MUI, with a default theme and a minimum of configuration (template provided by MUI)
- [Option 1.2](https://github.com/TOSIT-IO/tdp-ui/tree/6-interface-bootstrap): custom MUI, the PR was that one (template provided by MUI)
- [Option 2](https://github.com/TOSIT-IO/tdp-ui/tree/6-tailwind-template): Tailwind CSS (template provided by Tailwind UI and refactored)

To note :

- The simplest option is 1.
- Complexity increases a lot between options 1 and 1.2. We need to keep it in mind if we decide to go with 1 at first, and then customize it.
- Option 2 has a constant complexity as it is fully customized. It is manageable if the code is well organized.
- Option 2 could be further refactored. I've stopped so that the example is good enough and doesn't waste too much time.

**Comment by wdavidw** [comment]
React-bootstrap is not an option. I heard good feedbacks relative to Chakra UI.

Don't look at statistics. A project like MUI, related to google through the implementation of material design, distorts the numbers. This is a wrong evaluation approach.

Also, this quest of widget is not appropriate nor relevant in web design. In the end, we need control. It applies to web design just like it apply to TDP. Are you happy of spending your time working around Ansible limitations, I wouldn't. Widgets are no magic, they create an external abstraction layer which with time became a pain. Don't underestimate the power of Tailwind in that regard. It places styles where there are supposed to be in the most productive manner. Creates a simple layout representative of the complexity. Then, implement 3 POCs, one with Tailwind, one with MUI, one with a challenger (you can eliminate Ant and Bootstrap, that leave you one). Build, look at the generated file sizes, evaluate the quality of the docs, the community, its traction, search the occurrence of words on HN, look at GitHub trends and stats, share your feedbacks, thoroughly auto-critic the limitations of your evaluation, analysis your dependency locked-in, ... Finally move forward.
